### PR TITLE
Optimize the magicVariableToID function. 

### PR DESCRIPTION
**Pull request by vcvvvc**
The current time complexity is O(1).

Checked the magicVariables to ensure there are no omissions or mismatches.

### CODE REVIEW

**Comment by vcvvvc** [comment]
Trying fix ci/circleci: chk_coding_style error. 

~https://github.com/ethereum/solidity/pull/14899~
https://app.circleci.com/pipelines/github/ethereum/solidity/33122/workflows/4c8d08eb-e11d-4d48-b56e-9ae414ddd9b5/jobs/1488356

**Comment by vcvvvc** [comment]
I’m not sure about the specific coding style rules of the project, I might not be able to handle this issue. 💦

File: libsolidity/analysis/GlobalContext.cpp [hunk 1]
```diff
@@ -28,40 +28,33 @@
...
-	else if (_name == "sha3") return -23;
-	else if (_name == "suicide") return -24;
-	else if (_name == "super") return -25;
-	else if (_name == "tx") return -26;
-	else if (_name == "type") return -27;
-	else if (_name == "this") return -28;
-	else if (_name == "blobhash") return -29;
-	else
-		solAssert(false, "Unknown magic variable: \"" + _name + "\".");
+	static const std::unordered_map<std::string, int> magicVariables = {
```
Comment (nikola-matic) [inline libsolidity/analysis/GlobalContext.cpp]:
```suggestion
	static std::unordered_map<std::string, int> const magicVariables = {
```
Ideally, the entries should all be on separate lines as well, i.e.
```
{"abi", -1},
{"addmod", -2},
...
{"blobhash", -29}
```

File: libsolidity/analysis/GlobalContext.cpp [hunk 2]
```diff
@@ -28,40 +28,33 @@
...
+		{"revert", -19}, {"ripemd160", -20}, {"selfdestruct", -21}, {"sha256", -22},
+		{"sha3", -23}, {"suicide", -24}, {"super", -25}, {"tx", -26},
+		{"type", -27}, {"this", -28}, {"blobhash", -29}
+	};
+
+	auto it = magicVariables.find(_name);
+	if (it != magicVariables.end())
+	{
+		return it->second;
+	}
```
Comment (nikola-matic) [inline libsolidity/analysis/GlobalContext.cpp]:
```suggestion
	if (util::contains(magicVariables, _name))
		return magicVariables[_name];
```

**Review by nikola-matic — COMMENTED** [review]
Please squash the commits into one when you're done.

**Comment by vcvvvc** [comment]
```cpp
	return std::end(_t) != std::find(std::begin(_t), std::end(_t), _v);
```

`std::find` don't work with unordered_map, need to restore or create a new judgment function.

**Comment by nikola-matic** [comment]
> ```cpp
> 
> 	return std::end(_t) != std::find(std::begin(_t), std::end(_t), _v);
> 
> ```
> 
> 
> 
> `std::find` don't work with unordered_map, need to restore or create a new judgment function.
> 
> 

My bad, I wasn't aware `util::contains` works only on ordered containers. You can use 
```
if (magicVariables.count(_name)
    return ...
```

**Comment by vcvvvc** [comment]
> > ```c++
> > 	return std::end(_t) != std::find(std::begin(_t), std::end(_t), _v);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `std::find` don't work with unordered_map, need to restore or create a new judgment function.
> 
> My bad, I wasn't aware `util::contains` works only on ordered containers. You can use
> 
> ```
> if (magicVariables.count(_name)
>     return ...
> ```




> > ```c++
> > 	return std::end(_t) != std::find(std::begin(_t), std::end(_t), _v);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `std::find` don't work with unordered_map, need to restore or create a new judgment function.
> 
> My bad, I wasn't aware `util::contains` works only on ordered containers. You can use
> 
> ```
> if (magicVariables.count(_name)
>     return ...
> ```

```cpp
error: no viable overloaded operator[] for type 'const std::unordered_map<std::string, int>' (aka 'const unordered_map<basic_string<char, char_traits<char>, allocator<char>>, int>')
        return magicVariables[_name];
```
Because magicVariables it is const, we need to use `magicVariables.at()`
`at()` uses iterator, can this format be used?

```cpp
try {
        return magicVariables.at(_name);
    } catch (const std::out_of_range&) {
       solAssert(false, "Unknown magic variable: \"" + _name + "\".");
}
```

File: libsolidity/analysis/GlobalContext.cpp [hunk 3]
```diff
@@ -28,40 +28,48 @@
...
+		{"type", -27},
+        {"this", -28},
+        {"blobhash", -29}
+	};
+
+	try {
+        return magicVariables.at(_name);
+    } catch (const std::out_of_range&) {
+       solAssert(false, "Unknown magic variable: \"" + _name + "\".");
+	}
```
Comment (nikola-matic) [inline libsolidity/analysis/GlobalContext.cpp]:
```suggestion
	if (auto id = magicVariables.find(_name); id != magicVariables.end())
		return id->second;
	solAssert(false, "Unknown magic variable: \"" + _name + "\".");
```

**Comment by vcvvvc** [comment]
@nikola-matic  It's still a coding_style error.

**Comment by vcvvvc** [comment]
I tried it again.
The reason why the check failed seems to be the problem of code indentation formatting issues. There is no problem with the code and it should be able to be merge. @nikola-matic

**Review by r0qs — APPROVED** [review]

**Comment by nikola-matic** [comment]
Thanks @r0qs!

**Pull request merged**